### ISSUE TITLE
Mismatch between force_plot() output from DeepExplainer and actual model's output

### ISSUE THREAD

**Issue by tjiagoM** [shap/shap#1678]
Hi!

I have a Pytorch simple MLP model, which I'm trying to explain using `DeepExplainer`. 

I'm trying to check explanations on single samples but they don't correspond to the output that the model gives to that sample.

In the following printscreen I try to better explain what my issue is. The first two cells are just to show better how the variables that I'm using look like. In the third cell, you can see the output for my tensor `X_tmp` with 14 samples. Finally, in the last cell, I'm trying to explain the output for the first sample in X_tmp. However, as you can see, `force_plot` shows an output of `0.66` when the actual output from the model (in third cell) is ~`0.94`.

![Screenshot from 2020-12-21 18-22-47](https://user-images.githubusercontent.com/7031523/102809249-91db2100-43b9-11eb-831d-4f5147344193.png)

**Comment by aionfeehan** [comment]
I ran into this problem as well, and I've managed to replicate this and isolate somewhat (I think). I'm using pytorch=1.8.1 and shap=0.37 on linux.

I have a toy MLP defined as a class inheriting from torch.nn.Module, where the expected_value + shap_values != model(x), but the same model with no activations seems to work fine. Building the same model using torch.nn.Sequential() actually seems to work ok, even with activations, so @tjiagoM if this is still an issue and you can redefine the MLP using Sequential instead maybe try that.

I'm not exactly a master of either pytorch or shap, so I'm not quite sure where to start looking at what actually causes the behaviour. But I have uploaded the notebook in pdf here and in the link below, in case anyone more qualified has time to investigate.

[Shap Test.pdf](https://github.com/slundberg/shap/files/6251104/Shap.Test.pdf)


https://github.com/aionfeehan/aionfeehan/blob/main/Shap%20Test.ipynb

**Comment by aionfeehan** [comment]
I looked into this a bit and believe I found an answer.

If I understand correctly, DeepExplainer iterates through the different operations in the network essentially by accessing its ._modules, and adds handlers for gradients to each one. If it does not know how to handle an operation, or does not see the operation, it will be ignored (problem 1), and if you use the same module instance multiple times you may run into issues because it will get different handlers (problem 2, although I'm less clear on how this arises I think I have a solution). My notebook actually seems to demo these two different problems (completely unintentionally on my part), as I use torch.sigmoid for the sigmoid case (problem 1) and use the same activation instance across multiple layers (problem 2).

This is what I did that seems to have fixed the issue.

Problem 1 arises when you define non-linearities as passes through something that is not a torch.nn.Module, such as torch.sigmoid or anything from torch.nn.functional. Non-linearities seem to be handled properly if each activation function is defined as a Module that is an attribute of your network, like "self.hidden_activation = torch.nn.ReLU()"

Problem 2 seems to arise when you use the same module multiple times. In the forward pass in my network I use "h = self.activation(self.layer_1(x))" followed by "h = self.activation(layer_2(h))". If you change this by defining separate activation instances,  like "self.activation_1 = torch.nn.ReLU()" and "self.activation_2 = torch.nn.ReLU()", then redefine forward as "h = self.activation_1(self.layer_1(x))", "h = self.activation_2(self.layer_2(x))", you should get something that works.

I hope this is clear enough to be helpful. I'll update the notebook in the repo to show this.

**Comment by tjiagoM** [comment]
Hi @aionfeehan thank you so much for this, I'll have a go in a few weeks as I need to focus on something else, but indeed you seem to have discovered the source of the problem.

Curious to know whether @slundberg thinks this could be addressed in a future release of shap :)

**Comment by gianmarco-terrones** [comment]
As an alternative to DeepExplainer, you can try KernelExplainer. It works better for me.

**Comment by tjiagoM** [comment]
I know this came with a lot of delay, but I just wanted to confirm that the solution from @aionfeehan solved my issue, so it would be good to have this included in future releases of shap instead of manually having to change the model :) Thanks!
